### ISSUE TITLE
Bug: `no-redeclare` inconsistency for `var` in `catch` block

### ISSUE THREAD

**Issue by fasttime** [eslint/eslint#20422]
### Environment

Node version: v24.12.0
npm version: v11.6.2
Local ESLint version: v10.0.0-beta.0 (Currently used)
Global ESLint version: v1.0.0
Operating System: darwin 25.1.0

### What parser are you using?

Default (Espree)

### What did you do?

```js
/*eslint no-redeclare: "error"*/

try {
    undefined();
} catch (a) {
    var a = 10;
}

console.log(a);
```

### What did you expect to happen?

The rule should report the redeclaration of `a` inside the catch block.

### What actually happened?

No problem is being reported.

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tcmVkZWNsYXJlOiBcImVycm9yXCIqL1xuXG50cnkge1xuICAgIHVuZGVmaW5lZCgpO1xufSBjYXRjaCAoYSkge1xuICAgIHZhciBhID0gMTA7XG59XG5cbmNvbnNvbGUubG9nKGEpO1xuIiwib3B0aW9ucyI6eyJydWxlcyI6e30sImxhbmd1YWdlT3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319fX19

### Participation

- [x] I am willing to submit a pull request for this issue.

### Additional comments

I'm not entirely sure about this, so please double-check.

A `var` declaration inside a catch block that uses the same name as the catch parameter actually redeclares the same variable in the current scope rather than creating a new one. Because of that, it seems to me that the `no-redeclare` rule should report this case.

This is different from a `var` declaration inside a catch block with a different name than the catch parameter, which does declare a new variable in the outer scope, e.g.:

```js

try {
    undefined();
} catch (a) {
    var b = 10;
}

console.log(b);
```

**Comment by DMartens** [comment]
I think this is correct behaviour as this is not a redeclaration as
the `var` is hoisted. The `a` variable is declared in the upper `module` scope ([code-explorer](https://explorer.eslint.org/#eslint-explorer=IntcInN0YXRlXCI6e1widG9vbFwiOlwic2NvcGVcIixcImNvZGVcIjp7XCJqYXZhc2NyaXB0XCI6XCJ0cnkge1xcbiAgICB1bmRlZmluZWQoKTtcXG59IGNhdGNoIChhKSB7XFxuICAgIHZhciBhID0gMTA7XFxufVwiLFwianNvblwiOlwiLyoqXFxuICogVHlwZSBvciBwYXN0ZSBzb21lIEpTT04gaGVyZSB0byBsZWFybiBtb3JlIGFib3V0XFxuICogdGhlIHN0YXRpYyBhbmFseXNpcyB0aGF0IEVTTGludCBjYW4gZG8gZm9yIHlvdS5cXG4gKlxcbiAqIFRoZSB0YWJzIGFyZTpcXG4gKlxcbiAqIC0gQVNUIC0gVGhlIEFic3RyYWN0IFN5bnRheCBUcmVlIG9mIHRoZSBjb2RlLCB3aGljaCBjYW5cXG4gKiAgIGJlIHVzZWZ1bCB0byB1bmRlcnN0YW5kIHRoZSBzdHJ1Y3R1cmUgb2YgdGhlIGNvZGUuIFlvdVxcbiAqICAgY2FuIHZpZXcgdGhpcyBzdHJ1Y3R1cmUgYXMgSlNPTiBvciBpbiBhIHRyZWUgZm9ybWF0LlxcbiAqXFxuICogWW91IGNhbiBjaGFuZ2UgdGhlIHdheSB0aGF0IHRoZSBKU09OIGNvZGUgaXMgaW50ZXJwcmV0ZWRcXG4gKiBieSBjbGlja2luZyBcXFwiSlNPTlxcXCIgaW4gdGhlIGhlYWRlciBhbmQgc2VsZWN0aW5nIGRpZmZlcmVudFxcbiAqIG9wdGlvbnMuXFxuICpcXG4gKiBUaGlzIGV4YW1wbGUgaXMgaW4gSlNPTkMgbW9kZSwgd2hpY2ggYWxsb3dzIGNvbW1lbnRzLlxcbiAqL1xcblxcbntcXG4gICAgXFxcImtleTFcXFwiOiBbdHJ1ZSwgZmFsc2UsIG51bGxdLFxcbiAgICBcXFwia2V5MlxcXCI6IHtcXG4gICAgICAgIFxcXCJrZXkzXFxcIjogWzEsIDIsIFxcXCIzXFxcIiwgMWUxMCwgMWUtM11cXG4gICAgfVxcbn1cIixcIm1hcmtkb3duXCI6XCI8IS0tXFxuVHlwZSBvciBwYXN0ZSBzb21lIE1hcmtkb3duIGhlcmUgdG8gbGVhcm4gbW9yZSBhYm91dFxcbnRoZSBzdGF0aWMgYW5hbHlzaXMgdGhhdCBFU0xpbnQgY2FuIGRvIGZvciB5b3UuXFxuXFxuVGhlIHRhYnMgYXJlOlxcblxcbi0gQVNUIC0gVGhlIEFic3RyYWN0IFN5bnRheCBUcmVlIG9mIHRoZSBjb2RlLCB3aGljaCBjYW5cXG5iZSB1c2VmdWwgdG8gdW5kZXJzdGFuZCB0aGUgc3RydWN0dXJlIG9mIHRoZSBjb2RlLiBZb3VcXG5jYW4gdmlldyB0aGlzIHN0cnVjdHVyZSBhcyBKU09OIG9yIGluIGEgdHJlZSBmb3JtYXQuXFxuXFxuWW91IGNhbiBjaGFuZ2UgdGhlIHdheSB0aGF0IHRoZSBNYXJrZG93biBjb2RlIGlzIGludGVycHJldGVkXFxuYnkgY2xpY2tpbmcgXFxcIk1hcmtkb3duXFxcIiBpbiB0aGUgaGVhZGVyIGFuZCBzZWxlY3RpbmcgZGlmZmVyZW50XFxub3B0aW9ucy5cXG5cXG5UaGlzIGV4YW1wbGUgaXMgaW4gQ29tbW9uTWFyayBtb2RlLlxcbi0tPlxcblxcbiMgRVNMaW50IE1hcmtkb3duIEV4YW1wbGVcXG5cXG5UaGlzIGlzIGFuIGV4YW1wbGUgb2YgYSBNYXJrZG93biBmaWxlIHRoYXQgY2FuIGJlIHBhcnNlZFxcbmJ5IEVTTGludC4gTWFya2Rvd24gaXMgYSBzaW1wbGUgbWFya3VwIGxhbmd1YWdlIHRoYXQgaXNcXG5vZnRlbiB1c2VkIGZvciBkb2N1bWVudGF0aW9uLlxcblxcbiMjIEZlYXR1cmVzXFxuXFxuLSBNYWtlIHRoaW5ncyAqaXRhbGljKiwgKipib2xkKiosIG9yIGBjb2RlYFxcbi0gQ3JlYXRlIFtsaW5rc10oaHR0cHM6Ly9lc2xpbnQub3JnKVxcbi0gU3VwcG9ydHMgSFRNTCA8c3BhbiBzdHlsZT1cXFwiY29sb3I6IHJlZDtcXFwiPmVsZW1lbnRzPC9zcGFuPlxcbi0gTGlzdHNcXG4gIC0gTmVzdGVkIGxpc3RzXCIsXCJjc3NcIjpcIi8qKlxcbiAqIFR5cGUgb3IgcGFzdGUgc29tZSBDU1MgaGVyZSB0byBsZWFybiBtb3JlIGFib3V0XFxuICogdGhlIHN0YXRpYyBhbmFseXNpcyB0aGF0IEVTTGludCBjYW4gZG8gZm9yIHlvdS5cXG4gKlxcbiAqIFRoZSB0YWJzIGFyZTpcXG4gKlxcbiAqIC0gQVNUIC0gVGhlIEFic3RyYWN0IFN5bnRheCBUcmVlIG9mIHRoZSBjb2RlLCB3aGljaCBjYW5cXG4gKiAgIGJlIHVzZWZ1bCB0byB1bmRlcnN0YW5kIHRoZSBzdHJ1Y3R1cmUgb2YgdGhlIGNvZGUuIFlvdVxcbiAqICAgY2FuIHZpZXcgdGhpcyBzdHJ1Y3R1cmUgYXMgSlNPTiBvciBpbiBhIHRyZWUgZm9ybWF0LlxcbiAqXFxuICogWW91IGNhbiBjaGFuZ2UgdGhlIHdheSB0aGF0IHRoZSBDU1MgY29kZSBpcyBpbnRlcnByZXRlZFxcbiAqIGJ5IGNsaWNraW5nIFxcXCJDU1NcXFwiIGluIHRoZSBoZWFkZXIgYW5kIHNlbGVjdGluZyBkaWZmZXJlbnRcXG4gKiBvcHRpb25zLlxcbiAqL1xcblxcbkBpbXBvcnQgdXJsKCdodHRwczovL2ZvbnRzLmdvb2dsZWFwaXMuY29tL2NzczI%2FZmFtaWx5PVJvYm90bzp3Z2h0QDQwMDs3MDAmZGlzcGxheT1zd2FwJyk7XFxuXFxuYm9keSB7XFxuXFx0Zm9udC1mYW1pbHk6IHNhbnMtc2VyaWY7XFxufVxcblxcbmgxIHtcXG5cXHRjb2xvcjogIzMzMztcXG59XFxuXFxucCB7XFxuXFx0bWFyZ2luOiAxZW0gMDtcXG59XCIsXCJodG1sXCI6XCI8IURPQ1RZUEUgaHRtbD5cXG48IS0tXFxuVHlwZSBvciBwYXN0ZSBzb21lIEhUTUwgaGVyZSB0byBsZWFybiBtb3JlIGFib3V0XFxudGhlIHN0YXRpYyBhbmFseXNpcyB0aGF0IEVTTGludCBjYW4gZG8gZm9yIHlvdS5cXG5cXG5UaGUgdGFicyBhcmU6XFxuXFxuLSBBU1QgLSBUaGUgQWJzdHJhY3QgU3ludGF4IFRyZWUgb2YgdGhlIGNvZGUsIHdoaWNoIGNhblxcbmJlIHVzZWZ1bCB0byB1bmRlcnN0YW5kIHRoZSBzdHJ1Y3R1cmUgb2YgdGhlIGNvZGUuIFlvdVxcbmNhbiB2aWV3IHRoaXMgc3RydWN0dXJlIGFzIEpTT04gb3IgaW4gYSB0cmVlIGZvcm1hdC5cXG4tLT5cXG5cXG48aHRtbCBsYW5nPVxcXCJlblxcXCI%2BXFxuICAgIDxoZWFkPlxcbiAgICAgICAgPG1ldGEgY2hhcnNldD1cXFwiVVRGLThcXFwiPlxcbiAgICAgICAgPHRpdGxlPkhUTUw8L3RpdGxlPlxcbiAgICA8L2hlYWQ%2BXFxuICAgIDxib2R5PlxcbiAgICAgICAgPHA%2BVGV4dDwvcD5cXG4gICAgPC9ib2R5PlxcbjwvaHRtbD5cIn0sXCJsYW5ndWFnZVwiOlwiamF2YXNjcmlwdFwiLFwianNPcHRpb25zXCI6e1wicGFyc2VyXCI6XCJlc3ByZWVcIixcInNvdXJjZVR5cGVcIjpcIm1vZHVsZVwiLFwiZXNWZXJzaW9uXCI6XCJsYXRlc3RcIixcImlzSlNYXCI6dHJ1ZX0sXCJqc29uT3B0aW9uc1wiOntcImpzb25Nb2RlXCI6XCJqc29uY1wiLFwiYWxsb3dUcmFpbGluZ0NvbW1hc1wiOmZhbHNlfSxcImNzc09wdGlvbnNcIjp7XCJjc3NNb2RlXCI6XCJjc3NcIixcInRvbGVyYW50XCI6ZmFsc2V9LFwibWFya2Rvd25PcHRpb25zXCI6e1wibWFya2Rvd25Nb2RlXCI6XCJjb21tb25tYXJrXCIsXCJtYXJrZG93bkZyb250bWF0dGVyXCI6XCJvZmZcIn0sXCJodG1sT3B0aW9uc1wiOntcInRlbXBsYXRlRW5naW5lU3ludGF4XCI6XCJub25lXCIsXCJmcm9udG1hdHRlclwiOmZhbHNlfSxcIndyYXBcIjp0cnVlLFwidmlld01vZGVzXCI6e1wiYXN0Vmlld1wiOlwianNvblwiLFwic2NvcGVWaWV3XCI6XCJmbGF0XCIsXCJwYXRoVmlld1wiOlwiZ3JhcGhcIn0sXCJwYXRoSW5kZXhcIjp7XCJpbmRleFwiOjAsXCJpbmRleGVzXCI6MX0sXCJlc3F1ZXJ5U2VsZWN0b3JcIjp7XCJzZWxlY3RvclwiOlwiXCJ9fSxcInZlcnNpb25cIjowfSI%3D)).
A redeclaration can only happen in the same scope ([relevant function from eslint-scope](https://github.com/eslint/js/blob/5887c4d5919013127242b9c8822625198d8698b8/packages/eslint-scope/lib/scope.js#L349))

**Comment by aladdin-add** [comment]
I second @DMartens : in essence, the catchparameter acreates a block-scoped variable that is separate from the function-scoped (or global) variable declared with var. The assignment inside the catchblock modifies the local catchparameter, leaving the hoisted global variable unchanged. This is a specific example of variable shadowing, where the inner catchparameter ashadows the outer, hoisted variable `a`.

I think we should update the rule [[no-shadow]](https://eslint.org/docs/latest/rules/no-shadow) to report the issue (maybe behind an option?)

**Comment by fasttime** [comment]
> I think this is correct behaviour as this is not a redeclaration as the `var` is hoisted. The `a` variable is declared in the upper `module` scope ([code-explorer](https://explorer.eslint.org/#eslint-explorer=IntcInN0YXRlXCI6e1widG9vbFwiOlwic2NvcGVcIixcImNvZGVcIjp7XCJqYXZhc2NyaXB0XCI6XCJ0cnkge1xcbiAgICB1bmRlZmluZWQoKTtcXG59IGNhdGNoIChhKSB7XFxuICAgIHZhciBhID0gMTA7XFxufVwiLFwianNvblwiOlwiLyoqXFxuICogVHlwZSBvciBwYXN0ZSBzb21lIEpTT04gaGVyZSB0byBsZWFybiBtb3JlIGFib3V0XFxuICogdGhlIHN0YXRpYyBhbmFseXNpcyB0aGF0IEVTTGludCBjYW4gZG8gZm9yIHlvdS5cXG4gKlxcbiAqIFRoZSB0YWJzIGFyZTpcXG4gKlxcbiAqIC0gQVNUIC0gVGhlIEFic3RyYWN0IFN5bnRheCBUcmVlIG9mIHRoZSBjb2RlLCB3aGljaCBjYW5cXG4gKiAgIGJlIHVzZWZ1bCB0byB1bmRlcnN0YW5kIHRoZSBzdHJ1Y3R1cmUgb2YgdGhlIGNvZGUuIFlvdVxcbiAqICAgY2FuIHZpZXcgdGhpcyBzdHJ1Y3R1cmUgYXMgSlNPTiBvciBpbiBhIHRyZWUgZm9ybWF0LlxcbiAqXFxuICogWW91IGNhbiBjaGFuZ2UgdGhlIHdheSB0aGF0IHRoZSBKU09OIGNvZGUgaXMgaW50ZXJwcmV0ZWRcXG4gKiBieSBjbGlja2luZyBcXFwiSlNPTlxcXCIgaW4gdGhlIGhlYWRlciBhbmQgc2VsZWN0aW5nIGRpZmZlcmVudFxcbiAqIG9wdGlvbnMuXFxuICpcXG4gKiBUaGlzIGV4YW1wbGUgaXMgaW4gSlNPTkMgbW9kZSwgd2hpY2ggYWxsb3dzIGNvbW1lbnRzLlxcbiAqL1xcblxcbntcXG4gICAgXFxcImtleTFcXFwiOiBbdHJ1ZSwgZmFsc2UsIG51bGxdLFxcbiAgICBcXFwia2V5MlxcXCI6IHtcXG4gICAgICAgIFxcXCJrZXkzXFxcIjogWzEsIDIsIFxcXCIzXFxcIiwgMWUxMCwgMWUtM11cXG4gICAgfVxcbn1cIixcIm1hcmtkb3duXCI6XCI8IS0tXFxuVHlwZSBvciBwYXN0ZSBzb21lIE1hcmtkb3duIGhlcmUgdG8gbGVhcm4gbW9yZSBhYm91dFxcbnRoZSBzdGF0aWMgYW5hbHlzaXMgdGhhdCBFU0xpbnQgY2FuIGRvIGZvciB5b3UuXFxuXFxuVGhlIHRhYnMgYXJlOlxcblxcbi0gQVNUIC0gVGhlIEFic3RyYWN0IFN5bnRheCBUcmVlIG9mIHRoZSBjb2RlLCB3aGljaCBjYW5cXG5iZSB1c2VmdWwgdG8gdW5kZXJzdGFuZCB0aGUgc3RydWN0dXJlIG9mIHRoZSBjb2RlLiBZb3VcXG5jYW4gdmlldyB0aGlzIHN0cnVjdHVyZSBhcyBKU09OIG9yIGluIGEgdHJlZSBmb3JtYXQuXFxuXFxuWW91IGNhbiBjaGFuZ2UgdGhlIHdheSB0aGF0IHRoZSBNYXJrZG93biBjb2RlIGlzIGludGVycHJldGVkXFxuYnkgY2xpY2tpbmcgXFxcIk1hcmtkb3duXFxcIiBpbiB0aGUgaGVhZGVyIGFuZCBzZWxlY3RpbmcgZGlmZmVyZW50XFxub3B0aW9ucy5cXG5cXG5UaGlzIGV4YW1wbGUgaXMgaW4gQ29tbW9uTWFyayBtb2RlLlxcbi0tPlxcblxcbiMgRVNMaW50IE1hcmtkb3duIEV4YW1wbGVcXG5cXG5UaGlzIGlzIGFuIGV4YW1wbGUgb2YgYSBNYXJrZG93biBmaWxlIHRoYXQgY2FuIGJlIHBhcnNlZFxcbmJ5IEVTTGludC4gTWFya2Rvd24gaXMgYSBzaW1wbGUgbWFya3VwIGxhbmd1YWdlIHRoYXQgaXNcXG5vZnRlbiB1c2VkIGZvciBkb2N1bWVudGF0aW9uLlxcblxcbiMjIEZlYXR1cmVzXFxuXFxuLSBNYWtlIHRoaW5ncyAqaXRhbGljKiwgKipib2xkKiosIG9yIGBjb2RlYFxcbi0gQ3JlYXRlIFtsaW5rc10oaHR0cHM6Ly9lc2xpbnQub3JnKVxcbi0gU3VwcG9ydHMgSFRNTCA8c3BhbiBzdHlsZT1cXFwiY29sb3I6IHJlZDtcXFwiPmVsZW1lbnRzPC9zcGFuPlxcbi0gTGlzdHNcXG4gIC0gTmVzdGVkIGxpc3RzXCIsXCJjc3NcIjpcIi8qKlxcbiAqIFR5cGUgb3IgcGFzdGUgc29tZSBDU1MgaGVyZSB0byBsZWFybiBtb3JlIGFib3V0XFxuICogdGhlIHN0YXRpYyBhbmFseXNpcyB0aGF0IEVTTGludCBjYW4gZG8gZm9yIHlvdS5cXG4gKlxcbiAqIFRoZSB0YWJzIGFyZTpcXG4gKlxcbiAqIC0gQVNUIC0gVGhlIEFic3RyYWN0IFN5bnRheCBUcmVlIG9mIHRoZSBjb2RlLCB3aGljaCBjYW5cXG4gKiAgIGJlIHVzZWZ1bCB0byB1bmRlcnN0YW5kIHRoZSBzdHJ1Y3R1cmUgb2YgdGhlIGNvZGUuIFlvdVxcbiAqICAgY2FuIHZpZXcgdGhpcyBzdHJ1Y3R1cmUgYXMgSlNPTiBvciBpbiBhIHRyZWUgZm9ybWF0LlxcbiAqXFxuICogWW91IGNhbiBjaGFuZ2UgdGhlIHdheSB0aGF0IHRoZSBDU1MgY29kZSBpcyBpbnRlcnByZXRlZFxcbiAqIGJ5IGNsaWNraW5nIFxcXCJDU1NcXFwiIGluIHRoZSBoZWFkZXIgYW5kIHNlbGVjdGluZyBkaWZmZXJlbnRcXG4gKiBvcHRpb25zLlxcbiAqL1xcblxcbkBpbXBvcnQgdXJsKCdodHRwczovL2ZvbnRzLmdvb2dsZWFwaXMuY29tL2NzczI%2FZmFtaWx5PVJvYm90bzp3Z2h0QDQwMDs3MDAmZGlzcGxheT1zd2FwJyk7XFxuXFxuYm9keSB7XFxuXFx0Zm9udC1mYW1pbHk6IHNhbnMtc2VyaWY7XFxufVxcblxcbmgxIHtcXG5cXHRjb2xvcjogIzMzMztcXG59XFxuXFxucCB7XFxuXFx0bWFyZ2luOiAxZW0gMDtcXG59XCIsXCJodG1sXCI6XCI8IURPQ1RZUEUgaHRtbD5cXG48IS0tXFxuVHlwZSBvciBwYXN0ZSBzb21lIEhUTUwgaGVyZSB0byBsZWFybiBtb3JlIGFib3V0XFxudGhlIHN0YXRpYyBhbmFseXNpcyB0aGF0IEVTTGludCBjYW4gZG8gZm9yIHlvdS5cXG5cXG5UaGUgdGFicyBhcmU6XFxuXFxuLSBBU1QgLSBUaGUgQWJzdHJhY3QgU3ludGF4IFRyZWUgb2YgdGhlIGNvZGUsIHdoaWNoIGNhblxcbmJlIHVzZWZ1bCB0byB1bmRlcnN0YW5kIHRoZSBzdHJ1Y3R1cmUgb2YgdGhlIGNvZGUuIFlvdVxcbmNhbiB2aWV3IHRoaXMgc3RydWN0dXJlIGFzIEpTT04gb3IgaW4gYSB0cmVlIGZvcm1hdC5cXG4tLT5cXG5cXG48aHRtbCBsYW5nPVxcXCJlblxcXCI%2BXFxuICAgIDxoZWFkPlxcbiAgICAgICAgPG1ldGEgY2hhcnNldD1cXFwiVVRGLThcXFwiPlxcbiAgICAgICAgPHRpdGxlPkhUTUw8L3RpdGxlPlxcbiAgICA8L2hlYWQ%2BXFxuICAgIDxib2R5PlxcbiAgICAgICAgPHA%2BVGV4dDwvcD5cXG4gICAgPC9ib2R5PlxcbjwvaHRtbD5cIn0sXCJsYW5ndWFnZVwiOlwiamF2YXNjcmlwdFwiLFwianNPcHRpb25zXCI6e1wicGFyc2VyXCI6XCJlc3ByZWVcIixcInNvdXJjZVR5cGVcIjpcIm1vZHVsZVwiLFwiZXNWZXJzaW9uXCI6XCJsYXRlc3RcIixcImlzSlNYXCI6dHJ1ZX0sXCJqc29uT3B0aW9uc1wiOntcImpzb25Nb2RlXCI6XCJqc29uY1wiLFwiYWxsb3dUcmFpbGluZ0NvbW1hc1wiOmZhbHNlfSxcImNzc09wdGlvbnNcIjp7XCJjc3NNb2RlXCI6XCJjc3NcIixcInRvbGVyYW50XCI6ZmFsc2V9LFwibWFya2Rvd25PcHRpb25zXCI6e1wibWFya2Rvd25Nb2RlXCI6XCJjb21tb25tYXJrXCIsXCJtYXJrZG93bkZyb250bWF0dGVyXCI6XCJvZmZcIn0sXCJodG1sT3B0aW9uc1wiOntcInRlbXBsYXRlRW5naW5lU3ludGF4XCI6XCJub25lXCIsXCJmcm9udG1hdHRlclwiOmZhbHNlfSxcIndyYXBcIjp0cnVlLFwidmlld01vZGVzXCI6e1wiYXN0Vmlld1wiOlwianNvblwiLFwic2NvcGVWaWV3XCI6XCJmbGF0XCIsXCJwYXRoVmlld1wiOlwiZ3JhcGhcIn0sXCJwYXRoSW5kZXhcIjp7XCJpbmRleFwiOjAsXCJpbmRleGVzXCI6MX0sXCJlc3F1ZXJ5U2VsZWN0b3JcIjp7XCJzZWxlY3RvclwiOlwiXCJ9fSxcInZlcnNpb25cIjowfSI%3D)). A redeclaration can only happen in the same scope ([relevant function from eslint-scope](https://github.com/eslint/js/blob/5887c4d5919013127242b9c8822625198d8698b8/packages/eslint-scope/lib/scope.js#L349))

Yeah, this is a bit of an edge case in that the variable is hoisted but only retains its value inside the catch block. In theory, either `no-redeclare` or `no-shadow` should flag this (and `no-redeclare` feels like the more appropriate rule), but neither one does: ([Playgound link](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tcmVkZWNsYXJlOiBcImVycm9yXCIqL1xuLyplc2xpbnQgbm8tc2hhZG93OiBcImVycm9yXCIqL1xuXG50cnkge1xufSBjYXRjaCAoYSkge1xuICAgIHZhciBhO1xufVxuIiwib3B0aW9ucyI6eyJydWxlcyI6e30sImxhbmd1YWdlT3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319fX19)).

**Comment by fasttime** [comment]
> I think we should update the rule [[no-shadow]](https://eslint.org/docs/latest/rules/no-shadow) to report the issue (maybe behind an option?)

Works for me if that's the team's preference.

**Comment by mdjermanovic** [comment]
I agree with others about not updating the `no-redeclare` rule. `var a = 10` in the catch block declares variable in the upper scope but assigns to the catch variable, which is quite a weird behavior but technically not a redeclaration.

```js
try {
	undefined();
} catch (a) {
	var a = 10;
	console.log(a); // 10
}

console.log(a); // undefined; it would be a reference error without var a = 10;
```

**Comment by mdjermanovic** [comment]
The `no-shadow` rule does report shadowing in this case when the `"hoist"` option is set to `"all"`:

https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tc2hhZG93OiBbXCJlcnJvclwiLCB7IFwiaG9pc3RcIjogXCJhbGxcIiB9XSovXG5cbnRyeSB7XG4gICAgdW5kZWZpbmVkKCk7XG59IGNhdGNoIChhKSB7XG4gICAgdmFyIGEgPSAxMDtcbn1cblxuY29uc29sZS5sb2coYSk7XG4iLCJvcHRpb25zIjp7InJ1bGVzIjp7fSwibGFuZ3VhZ2VPcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fX19fX0=

**Comment by aladdin-add** [comment]
> The `no-shadow` rule does report shadowing in this case when the `"hoist"` option is set to `"all"`:
> 
> https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tc2hhZG93OiBbXCJlcnJvclwiLCB7IFwiaG9pc3RcIjogXCJhbGxcIiB9XSovXG5cbnRyeSB7XG4gICAgdW5kZWZpbmVkKCk7XG59IGNhdGNoIChhKSB7XG4gICAgdmFyIGEgPSAxMDtcbn1cblxuY29uc29sZS5sb2coYSk7XG4iLCJvcHRpb25zIjp7InJ1bGVzIjp7fSwibGFuZ3VhZ2VPcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fX19fX0=

right, seems the issue can be closed - imho, it makes more sense to be enabled by default, but it's a breaking change and seems to be beyond the scope of this issue. 😄

**Comment by fasttime** [comment]
> The `no-shadow` rule does report shadowing in this case when the `"hoist"` option is set to `"all"`:
> https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tc2hhZG93OiBbXCJlcnJvclwiLCB7IFwiaG9pc3RcIjogXCJhbGxcIiB9XSovXG5cbnRyeSB7XG4gICAgdW5kZWZpbmVkKCk7XG59IGNhdGNoIChhKSB7XG4gICAgdmFyIGEgPSAxMDtcbn1cblxuY29uc29sZS5sb2coYSk7XG4iLCJvcHRpb25zIjp7InJ1bGVzIjp7fSwibGFuZ3VhZ2VPcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fX19fX0=

Ah, I see now, thanks!

> right, seems the issue can be closed - imho, it makes more sense to be enabled by default, but it's a breaking change and seems to be beyond the scope of this issue. 😄

Yes, let's close this then. I can add unit tests to both `no-redeclare` and `no-shadow` to safeguard the current behavior.
